feat(exchange-provider): exchange provider

Daryl-L opened the pull request:
Feature #80

Daryl-L committed:
feat(exchange-manager): finish exchange manager

Daryl-L committed:
feat(exchange): finish rpc handler

homura commented:
@Daryl-L the CI is failed, can you please fix it

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,117 @@
+    const cells = (
+      await indexer.getCells({
+        script_type: 'lock',
+        script: helpers.parseAddress(this.providerAddress),
```
we can use the `provider.parseToAddress` instead of Lumos here, since the issuer server dependend on ckit, helper may not init and cause parse address error

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,117 @@
+    this.initiated = false;
+  }
+
+  shouldInitiated(): void {
```
```suggestion
  assertInitialted(): void {
```

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,117 @@
+  private async refreshCells(): Promise<void> {
+    this.shouldInitiated();
+
+    const indexer = new Indexer(this.context.ckitProvider.mercuryUrl, this.context.ckitProvider.rpcUrl);
```
```ts
  this.context.ckitProvider.collectUdtCells(address, sudtTypeScript, U128_MAX)
```

can be refactored to `collectUdtCells` here.

`indexer.getCells({script_type, script})` is also possible to collect cells from other sudt, which is not safe to collect

Daryl-L replied:
> ```ts
>   this.context.ckitProvider.collectUdtCells(address, sudtTypeScript, U128_MAX)
> ```
> 
> can be refactored to `collectUdtCells` here.
> 
> `indexer.getCells({script_type, script})` is also possible to collect cells from other sudt, which is not safe to collect

The only thing I can get from `collectUdtCells` is outpoint, so I try to add the sudt script to the filter, to be sure that only the certain sudt cells can be collected.

homura commented on packages/app-server-issuer/src/exchange-provider/config.ts:
```diff
@@ -0,0 +1,6 @@
+interface Config {
+  cellAmount: number;
```
```suggestion
  exchangeCellCount: number;
```

homura commented on packages/app-server-issuer/src/index.ts:
```diff
@@ -41,7 +48,7 @@ async function initContext(): Promise<ServerContext> {
     },
   });
-  return { ckitProvider, txSigner, rcHelper };
+  return { ckitProvider, txSigner, rcHelper, privateKey };
```
we can use a exchangeSigner here instead of the privateKey

homura commented on packages/app-server-issuer/src/exchange-provider/config.ts:
```diff
@@ -0,0 +1,6 @@
+  cellAmount: number;
+  sudtArgs: string;
+  initCapacity: number;
+  sudtExchangeRate: number;
```
using rate may encounter issues with the precision of floating point numbers, we can use a config like this


```sh
# 1 sUDT = 100 CKB
# the 0x... is args of the sUDT
# the decimals with the sUDT is 18
EXCHANGE_SUDT_AMOUNT_0x12345678=1000000000000000000 # 1e18
EXCHANGE_CKB_AMOUNT_0x12345678=10000000000 # 100e8
```

```ts
// 1 sUDT = 100 CKB
// assume the decimals of the sUDT is 18

const exchangeCkbAmount = BigNumber.from(process.env.EXCHANGE_CKB_AMOUNT_0x12345678)
const exchangeSudtAmount = BigNumber.from(process.env.EXCHANGE_SUDT_AMOUNT_0x12345678)

const exchangeableCkbAmount = BigNumber.from(readyForExchangeSudtAmount)
  .mul(exchangeCkbAmount)
  .div(exchangeSudtAmount)
```

and we can define the exchange price in a `.env` like this

Daryl-L replied:
> using rate may encounter issues with the precision of floating point numbers, we can use a config like this
> 
> ```shell
> # 1 sUDT = 100 CKB
> # the 0x... is args of the sUDT
> # the decimals with the sUDT is 18
> EXCHANGE_SUDT_AMOUNT_0x12345678=1000000000000000000 # 1e18
> EXCHANGE_CKB_AMOUNT_0x12345678=10000000000 # 100e8
> ```
> 
> ```ts
> // 1 sUDT = 100 CKB
> // assume the decimals of the sUDT is 18
> 
> const exchangeCkbAmount = BigNumber.from(process.env.EXCHANGE_CKB_AMOUNT_0x12345678)
> const exchangeSudtAmount = BigNumber.from(process.env.EXCHANGE_SUDT_AMOUNT_0x12345678)
> 
> const exchangeableCkbAmount = BigNumber.from(readyForExchangeSudtAmount)
>   .mul(exchangeCkbAmount)
>   .div(exchangeSudtAmount)
> ```
> 
> and we can define the exchange price in a `.env` like this

good idea!

Daryl-L committed:
fix(manager): use parse address in ckit provider instead of which in lumos

Daryl-L committed:
fix(manager): udt script must be one of the filters to get cells

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -38,6 +38,13 @@ export class ExchangeProviderManager {
         script_type: 'lock',
         script: this.context.ckitProvider.parseToScript(this.providerAddress),
+        filter: {
+          script: {
+            hash_type: this.context.ckitProvider.getScriptConfig('SUDT').HASH_TYPE,
+            code_hash: this.context.ckitProvider.getScriptConfig('SUDT').CODE_HASH,
+            args: this.config.sudtArgs,
+          },
```
can be refactored to
```ts
provider.newScript('SUDT', this.config.sudtArgs)
```

Daryl-L committed:
refactor(manager): config for exchange rate, use ckb amount and sudt amount to compute

Daryl-L committed:
refactor(deposit): initiate an exchange signer instead of using private directly

Daryl-L committed:
style(exchange-manager): some code style changed

1. use provider to initiate sUDT script instead of initiating it manually
2. change shouldInitiated name to assertInitiated
3. use exchangeCellCount instead of cellAmount

Daryl-L committed:
style(deposit): useless import removed

Daryl-L committed:
fix(deposit): SerializedTx imported in type, not in rpc

Daryl-L committed:
fix(deposit): use unknown type

Daryl-L committed:
fix(manager): use lumos and ethers in package.json

homura commented on package.json:
```diff
@@ -45,5 +45,9 @@
+  },
+  "dependencies": {
+    "@ckb-lumos/lumos": "^0.17.0-rc9",
+    "ethers": "^5.6.9"
```
why we need ethers here?

homura commented on packages/app-server-issuer/src/exchange-provider/config.ts:
```diff
@@ -0,0 +1,13 @@
+import { BigNumber } from 'ethers';
```
lumos has exposed a big integer library `@ckb-lumos/bi`, it is more light-weight, we can use it

Daryl-L committed:
refactor(deposit): use BI from lumos instead of BigNumber from ethers

Daryl-L committed:
feat(deposit): update version of ckit dependency

Daryl-L committed:
feat(deposit): remove ethers dependency

Daryl-L committed:
feat(deposit): add handler for deposit tx generation

Daryl-L committed:
fix(manager): wrong spelling for env

Daryl-L committed:
fix(manager): should use secp256k1 address to hold cells

Daryl-L committed:
fix(manager): wrong spelling for env

Daryl-L committed:
feat(docker): new env paramaters

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -72,27 +71,31 @@ export class ExchangeProviderManager {
 
   exchangeAmount(sudtAmount: HexString): BigNumber {
-    return BigNumber.from(sudtAmount).mul(this.config.exchange.CKB).div(this.config.exchange.sUDT);
+    return BigNumber.from(sudtAmount)
+      .mul(this.config.exchange.CKB)
+      .div(this.config.exchange.sUDT)
+      .mul(BigNumber.from(10).pow(9));
```
maybe there are some mistakes here,

1. we should multiply first and divide later to reduce the error caused by floating number

```
x / y = x1 / y1

# mul first and divide later
x = (x1 * y) / y1
```
2. all exposed RPC methods do not carry floating point values, for example CKB uses `shanon` as the unit, i think the `.mul(BigNumber.from(10).pow(9))` here is a mistake

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -72,27 +71,31 @@ export class ExchangeProviderManager {
-    await this.context.ckitProvider.waitForTransactionCommitted(hash);
+    try {
+      const unsignedTx = await builder.build();
+      const tx = await this.context.txSigner.seal(unsignedTx);
```
```suggestion
      const tx = await this.context.txSigner.partialSeal(unsignedTx);
```

Daryl-L replied:
> 

For manager, cells creation could only be signed by acp signer, no other one needs to provider extra cells for input, I think. Maybe you mean in handler I should change the code like this, is it right?

homura replied:
OK i see

Daryl-L committed:
fix(deposit): unsigned tx should be partial sealed

Daryl-L committed:
fix(manager): the sudt cell capacity should be included in needed capacity

Daryl-L committed:
feat(manager): CKB in config should be shannon

Daryl-L committed:
style(manager): remove useless import

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,129 @@
+          lock: this.context.ckitProvider.parseToScript(this.context.exchangeSigner!.getAddress()),
+          type: this.context.ckitProvider.newScript('SUDT', this.config.sudtArgs),
+        },
+        data: '0x10',
```
```suggestion
        data: Uint128.pack(0),
```

felicityin commented on packages/app-server-issuer/src/rpc-server/handler.ts:
```diff
@@ -68,6 +74,33 @@ export class IssuerRpcHandler implements rpc.IssuerRpc {
+      },
+      this.context.ckitProvider,
+    );
+    const unsignedTx = await builder.build();
```
This line may throw an exception. And the exception should be returned.

Daryl-L committed:
feat(manager): use codec package in lumos to fill the sUDT cell data

Daryl-L committed:
feat(manager): add config for limitation of exchangeable CKB amount

Daryl-L committed:
fix(manager): fix the initiation of exchange provider cell

1. the amount of cells, should be same with configuration, not for all
2. avoid the deduplicate the same recipient cell

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,17 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+
+## exchange configs
+# initiated amount of cells for exchange
+CELL_AMOUNT=100
```
```suggestion
EXCHANGE_CELL_COUNT=100
```

the `amount` here is confused, maybe use the `count` here is more appropriate

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,17 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+# initiated amount of cells for exchange
+CELL_AMOUNT=100
+# the capacity to initiate for exchange cells
+INIT_CAPACITY=10000
```
```suggestion
EXCHANGE_INIT_CAPACITY=1000000000000 # 10,000 CKB
```

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,17 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+# the capacity to initiate for exchange cells
+INIT_CAPACITY=10000
+# sUDT script args
+SDUT_ARGS=0x...
```
```suggestion
EXCHANGE_SDUT_ARGS=0x...
```

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,17 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+# the decimals with the sUDT is 18
+EXCHANGE_SUDT_AMOUNT=1000000000000000000 # 1e18
+EXCHANGE_CKB_AMOUNT=10000000000 # 100e8
+EXCHANGE_CKB_LIMIT=10000000000000000
```
```suggestion
EXCHANGE_CKB_LIMIT_EACH_TRANSACTION=10000000000000000
```

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,136 @@
+    const needCapacity = this.exchangeAmount(sudtAmount).add(
+      helpers.minimalCellCapacity({
+        cell_output: {
+          capacity: '0x10',
```
```suggestion
          capacity: '0x0',
```

Daryl-L replied:
The function `minimalCellCapicity` only accepts from `0x10` to `0xff` for capicity, or it will throw an error. That's why I use this.

homura commented on packages/commons/src/interfaces/rpc.ts:
```diff
@@ -128,3 +128,15 @@ export interface DisableClaimSecretPayload {
+}
+
+export interface GenerateDepositToGodWokenResponse {
+  transaction: unknown;
```
the `unknown` here is too magic for an exposed API, we should define the type of `SerializedTransaction`

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,136 @@
+  private async refreshCells(): Promise<void> {
+    this.assertInitiated();
+
+    const indexer = new Indexer(this.context.ckitProvider.mercuryUrl, this.context.ckitProvider.rpcUrl);
+    const cells = (
+      await indexer.getCells({
+        script_type: 'lock',
+        script: this.context.ckitProvider.parseToScript(this.context.exchangeSigner!.getAddress()),
+        filter: {
+          script: this.context.ckitProvider.newScript('SUDT', this.config.sudtArgs),
+        },
+      })
+    ).objects;
```
maybe it is unnecessary here, the [collectCells](https://github.com/nervosnetwork/ckit/blob/develop/packages/ckit/src/providers/mercury/MercuryProvider.ts#L155) is more friendly

```ts
await provider.collectCells(script_type: 'lock',
	        script: this.context.ckitProvider.parseToScript(this.context.exchangeSigner!.getAddress()),
	        filter: {
	          script: this.context.ckitProvider.newScript('SUDT', this.config.sudtArgs),
	        },
	      })
	    ))

```

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,136 @@
+      const hash = await this.context.ckitProvider.sendTransaction(tx);
+      await this.context.ckitProvider.waitForTransactionCommitted(hash);
+    } catch (e) {
+      console.log(e);
```
```suggestion
      throw new Error(`init provider cell failed, please try again later`);
```

Daryl-L replied:
Here I want to print error to the screen, so I will use `logger` instead of using `console.log` directly.

Daryl-L committed:
feat(deposit): use serializer in ckit to serialize partial signed tx for response

Daryl-L committed:
feat(manager): use collectCells in ckit instead of using indexer in lumos

Daryl-L committed:
feat(manager): use log instead of using console.log

Daryl-L committed:
style(manager): env property rename

Daryl-L committed:
feat(exchange): private key should be provided for exchange provider

Daryl-L committed:
fix(deposit): redefine the serialized tx type

Daryl-L committed:
fix(manager): fix an error in refreshCell call

If a request from user, ask to exchange some ckb, and the ckb is not enough
for user to exchange, the manager will refresh cells. But if the configuration
 of exchange cell amount is 1, that means the manager will not refresh the cells
since the amount reaches, even though the capacity does not. So this commit is to solve
this problem. To check the cell amount as well as the total capacity.

Daryl-L committed:
feat(builder): submodule version updated for test

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,18 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+
+## exchange configs
+# initiated amount of cells for exchange
+EXCHANGE_CELL_AMOUNT=100
```
```suggestion
# initiated count of cells for exchange
EXCHANGE_CELL_COUNT=100
```

`count` is more appropriate here than amount, the word amount more like an asset amount

homura commented on docker/.env.aggron:
```diff
@@ -48,3 +48,18 @@ REACT_APP_NERVOS_EXPLORER_URL=https://explorer.nervos.org/aggron
+# initiated amount of cells for exchange
+EXCHANGE_CELL_AMOUNT=100
+# the capacity to initiate for exchange cells
+EXCHANGE_INIT_CAPACITY=10000
```
```suggestion
# the capacity to initiate for exchange cells
EXCHANGE_INIT_CAPACITY=1000000000000 #10000e8
```

we should keep using the same precision in a config file

homura commented on package.json:
```diff
@@ -45,5 +45,8 @@
     "buffer": "^6.0.3"
+  },
+  "dependencies": {
+    "@ckb-lumos/lumos": "^0.17.0-rc9"
```
maybe we can upgrade to 0.18?

Daryl-L replied:
> maybe we can upgrade to 0.18?

I will have a try

homura replied:
found the ckit using 0.17.0-rc9, maybe we can upgrade lumos in the future after ckit has upgraded lumos

Daryl-L replied:
> maybe we can upgrade to 0.18?

If I did this, the CI would fail

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,167 @@
+export class ExchangeProviderManager {
+  private cells: Cell[] = [];
+  private initiated: boolean;
+  private config!: Config;
+  private context!: ServerContext;
```
```suggestion
  private config: Config;
  private context: ServerContext;
```

can we remove `!` here? we can throw an error when constructing failed

Daryl-L replied:
I made it a singleton, the configs will not initiated in constructor. Just like this.
```typescript
ExchangeProviderManager.getInstance().initiateConfig(config);
```
So that we can call the `getInstance` without any parameters.

homura commented on packages/app-server-issuer/src/types/index.ts:
```diff
@@ -1,9 +1,14 @@
   ckitProvider: CkitProvider;
   txSigner: InstanceType<typeof internal['Secp256k1Signer']>;
   rcHelper: RcSupplyLockHelper;
+  exchangeSigners?: Partial<
+    {
+      [field in CkitConfigKeys]: InstanceType<typeof internal['Secp256k1Signer']>;
+    }
+  >;
```
```suggestion
  exchangeSigner: internal.Secp256k1Signer;
```

why we need multi kinds of signer for `exchangeSigner`?

Daryl-L replied:
Because that the acp builder in ckit doesn't allow the address of sender and reciever to be the same, for convenient, here I use multiple lockscripts to do this.

homura commented on packages/app-server-issuer/src/exchange-provider/ExchangeProviderManager.ts:
```diff
@@ -0,0 +1,167 @@
+    await this.refreshCells();
+  }
+
+  exchangeAmount(sudtAmount: HexString): BigNumber {
```
```suggestion
  calcExchangableCkbAmount(sudtAmount: HexString): BigNumber {
```

homura commented on packages/app-server-issuer/src/exchange-provider/config.ts:
```diff
@@ -0,0 +1,14 @@
+export interface ExchangeConfig {
+  sUDT: BigNumber;
+  CKB: BigNumber;
+  limitation: BigNumber;
```
```suggestion
  pairedBaseSudtPrice: BigNumber;
  pairedQuoteCkbPrice: BigNumber;
  maxCapacityPerExchange: BigNumber;
```

Daryl-L committed:
style(exchange): some configuration style changed

homura commented on .gitmodules:
```diff
@@ -1,3 +1,4 @@
 [submodule "ckit"]
 	path = ckit
-	url = https://github.com/nervosnetwork/ckit.git
+	url = https://github.com/felicityin/ckit.git
+	branch = fix-exchange-builder
```
can be upgraded to latest ckit

Daryl-L committed:
feat(exchange): update submodule ckit

Daryl-L committed:
feat(exchange): use tx signer to create cell for exchange provider
